Add support for MySQL (#7538)
keystonejs/keystone: docs/pages/docs/guides/choosing-a-database.mdx

import { Markdown } from '../../../components/Markdown';
import { Well } from '../../../components/primitives/Well';
import { RelatedContent } from '../../../components/RelatedContent';

# Choosing the right Database

Keystone supports Postgres, MySQL and SQLite database engines. This guide highlights the differences between these engines to help you choose the right one for your project.

> !> **Note:** SQLite is not recommended in production unless for readonly senarios like the [embedded Keystone](../walkthroughs/embedded-mode-with-sqlite-nextjs) example

Author: ```suggestion
!> **Note:** SQLite is not recommended in production except for scenarios like the [embedded Keystone](../walkthroughs/embedded-mode-with-sqlite-nextjs) example
```

## Case Sensitivity

> By default, Postgres is case sensitive, while MySQL and SQLite are not case sensitive. This means when filtering and sorting the results will differ depending on the database type and collation used.

Reviewer: SQLite is only case insensitive for `contains`, `startsWith` and `endsWith`

Author: Thanks @mitchellhamilton, just updated

> When using `StringFilter` in a GraphQL query the `mode` defaults to that of the database - case sensitive for Postgres, and collation dependent for MySQL. You can make a query case insensitive by using `mode: insensitive`, there is no way to make a `StringQuery` case sensitive if your database is not case sensitive.

Reviewer: `mode` only exists on Postgres. Also, `StringFilter` & `StringQuery` -> string filter imo

Author: Thanks @mitchellhamilton, I just rewrote this section a bit.

import { Markdown } from '../../../components/Markdown';
import { Well } from '../../../components/primitives/Well';
import { RelatedContent } from '../../../components/RelatedContent';

# Choosing the right Database

> Keystone supports Postgres, MySQL and SQLite as database engines for your backend. There might be different reasons why you would choose one over the other.

Reviewer 2: Should we link to something for `Postgres`, `MySQL`, `SQLite`?

Perhaps other parts of our documentation that show how to configure each? 🤷

Author: Added some links...

> Keystone supports Postgres, MySQL and SQLite as database engines for your backend. There might be different reasons why you would choose one over the other.

Reviewer 2: ```suggestion
Keystone supports Postgres, MySQL and SQLite database engines. This guide highlights the differences between these engines to help you choose the right one for your project.
```

> This page seeks to clarify some of the differences in the databases supported by Keystone.

Reviewer 2: ```suggestion
```

> !> **Note:** SQLite is not recommended in production unless for readonly senarios like the [embedded Keystone](../walkthroughs/embedded-mode-with-sqlite-nextjs) example

Reviewer 2: Is the use of `readonly` consistent with rest of docs?

See the [Case Sensitivity Prisma Docs](https://www.prisma.io/docs/concepts/components/prisma-client/case-sensitivity) for more information.

## Default Types

Prisma has different default types for each database used, for example, the Keystone `text` field type uses the Prisma type of `String`,
for Postgres, Prisma uses the `text` column type, and for MySQL, it uses the `varchar(191)` column type. For more details on the `field` type differences
see [Primsa Schema Reference](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#model-field-scalar-types).

The Keystone `text` field supports the `db.nativeType` option, allowing you to override this - [Fields API](../apis/fields). Over time this option will be added to other field types.

## Prisma Resources

- [Database Features](https://www.prisma.io/docs/reference/database-reference/database-features)
- [Supported Databases](https://www.prisma.io/docs/reference/database-reference/supported-databases)

> !> **Note**: Keystone currently supports Postgres, MySQL, SQLite

Reviewer 2: Not sure we need this due to the presence in the top of page callout.

If you'd like to keep it, suggest including `Prisma` somewhere to situate it in the Prisma H2 in a more obvious way.

Author: Thanks @raveling I have just rewritten this section

## Related resources

<RelatedContent>
  <Well
    heading="Fields API"
    href="../apis/fields"
    >
    Documentation for the Fields API
  </Well>
  <Well
    heading="DB Config API"
    href="../apis/config#db"
    >
    Documentation for the DB Config API 
  </Well>
</RelatedContent>

> export default ({ children }) => <Markdown description="Learn which database is right for your project">{children}</Markdown>;

Reviewer 2: ```suggestion
export default ({ children }) => <Markdown description="How to choose the right database for your Keystone project">{children}</Markdown>;
```
